be/src/exec/pipeline/sort/sort_context.cpp

```diff
@@ -2,69 +2,111 @@
-    _offset -= skipped_rows;
-    if (res != nullptr) {
-        RETURN_IF_ERROR(res->downgrade());
-    }
+    while (_required_rows > 0 && !_merger.is_eos()) {
+        if (_current_chunk.empty()) {
+            auto chunk = _merger.try_get_next();
+            // Input cursor maye short circuit
+            if (!chunk) {
+                _required_rows = 0;
+                return nullptr;
+            }
+            _current_chunk.reset(std::move(chunk));
+        }
 
-    if (run.empty()) {
-        _merged_runs.pop_front();
+        // Skip some rows before return it
+        if (_offset > 0) {
+            _offset -= _current_chunk.skip(_offset);
+            if (_current_chunk.empty()) {
+                continue;
+            }
+        }
+
+        // Return required rows and cutoff the big chunk
+        size_t required_rows = std::min<size_t>(_required_rows, _state->chunk_size());
+        ChunkPtr res = _current_chunk.cutoff(required_rows);
+        _required_rows -= res->num_rows();
+        return res;
     }
-    return res;
+    return nullptr;
 }
 
-Status SortContext::_merge_inputs() {
-    int64_t total_rows = _total_rows.load(std::memory_order_relaxed);
+Status SortContext::_init_merger() {
+    if (_merger_inited) {
+        return Status::OK();
+    }
 
     // Keep all the data if topn type is RANK or DENSE_RANK
-    int64_t require_rows = total_rows;
+    _required_rows = _total_rows;
     if (_topn_type == TTopNType::ROW_NUMBER) {
-        require_rows = ((_limit < 0) ? total_rows : std::min(_limit + _offset, total_rows));
+        _required_rows = ((_limit < 0) ? _total_rows.load() : std::min<int64_t>(_limit + _offset, _total_rows));
     }
 
     std::vector<SortedRuns> partial_sorted_runs;
     for (int i = 0; i < _num_partition_sinkers; ++i) {
-        auto& partition_sorter = _chunks_sorter_partions[i];
+        auto& partition_sorter = _chunks_sorter_partitions[i];
         partial_sorted_runs.push_back(partition_sorter->get_sorted_runs());
     }
+    _partial_cursors.reserve(_num_partition_sinkers);
+    for (int i = 0; i < _num_partition_sinkers; i++) {
+        vectorized::ChunkProvider provider = [i, this](vectorized::ChunkUniquePtr* out_chunk, bool* eos) -> bool {
+            // data ready
+            if (out_chunk == nullptr || eos == nullptr) {
+                return true;
+            }
+            auto& partition_sorter = _chunks_sorter_partitions[i];
+            ChunkPtr chunk;
+            Status st = partition_sorter->get_next(&chunk, eos);
+            if (!st.ok() || *eos || chunk == nullptr) {
+                return false;
+            }
+            *out_chunk = chunk->clone_unique();
+            return true;
+        };
+        _partial_cursors.push_back(
+                std::make_unique<vectorized::SimpleChunkSortCursor>(std::move(provider), &_sort_exprs));
+    }
+
+    RETURN_IF_ERROR(_merger.init(_sort_desc, std::move(_partial_cursors)));
```

Comment:
Do you have test MergeCursorsCascade cost time with `blocking sort` in different scales?

Reply:
I doubt `MergeCursorsCascade ` maybe  compare to `partial_sort` if scale is not too big.

Reply:
The backboned merge algorithm is almost same, both are cascade multi-pass two-run merge. The only difference is the first is blocking, the second is streaming.